tensordict/prototype/tensorclass.py

```diff
@@ -165,65 +165,45 @@ def _init_wrapper(init, expected_keys):
-                else:
-                    default = field.default
-                if default not in (None, dataclasses.MISSING):
-                    kwargs.setdefault(key, default)
-
-            missing_params = [p for p in required_params if p not in kwargs]
-            if missing_params:
-                n_missing = len(missing_params)
-                raise TypeError(
-                    f"{self.__class__.__name__}.__init__() missing {n_missing} "
-                    f"required positional argument{'' if n_missing == 1 else 's'}: "
-                    f"""{", ".join(f"'{name}'" for name in missing_params)}"""
-                )
+    def wrapper(self, *args, batch_size, device=None, **kwargs):
+        for value, key in zip(args, self.__dataclass_fields__):
+            if key in kwargs:
+                raise ValueError(f"The key {key} is already set in kwargs")
+            kwargs[key] = value
+
+        for key, field in self.__dataclass_fields__.items():
+            if field.default_factory is not dataclasses.MISSING:
+                default = field.default_factory()
+            else:
+                default = field.default
+            if default not in (None, dataclasses.MISSING):
+                kwargs.setdefault(key, default)
 
-            self.tensordict = TensorDict({}, batch_size=batch_size, device=device)
-            init(
-                self, **{key: _get_typed_value(value) for key, value in kwargs.items()}
+        missing_params = [p for p in required_params if p not in kwargs]
+        if missing_params:
+            n_missing = len(missing_params)
+            raise TypeError(
+                f"{self.__class__.__name__}.__init__() missing {n_missing} "
+                f"required positional argument{'' if n_missing == 1 else 's'}: "
+                f"""{", ".join(f"'{name}'" for name in missing_params)}"""
             )
 
+        self.tensordict = TensorDict({}, batch_size=batch_size, device=device)
+        init(self, **{key: _get_typed_value(value) for key, value in kwargs.items()})
+
     new_params = [
-        inspect.Parameter(
-            "batch_size", inspect.Parameter.POSITIONAL_OR_KEYWORD, default=None
-        ),
-        inspect.Parameter(
-            "device", inspect.Parameter.POSITIONAL_OR_KEYWORD, default=None
-        ),
-        inspect.Parameter(
-            "_tensordict", inspect.Parameter.POSITIONAL_OR_KEYWORD, default=None
-        ),
+        inspect.Parameter("batch_size", inspect.Parameter.KEYWORD_ONLY),
+        inspect.Parameter("device", inspect.Parameter.KEYWORD_ONLY, default=None),
     ]
     wrapper.__signature__ = init_sig.replace(parameters=params + new_params)
 
     return wrapper
 
 
-def _build_from_tensordict(cls, tensordict):
-    return cls(_tensordict=tensordict)
+def _from_tensordict(cls, tensordict):
+    tc = cls(**tensordict, batch_size=tensordict.batch_size)
+    tc.__dict__["tensordict"] = tensordict
+    return tc
```

Comment:
```suggestion
def _from_tensordict(cls, tensordict, copy=True):
    #TODO assert tensordict is tensordict
    if copy:
        tc = cls(**tensordict, batch_size=tensordict.batch_size)
    else:
            input_dict = {key: None for key in tensordict.keys()}
            init(self, **input_dict, batch_size=tensordict.batch_size)
            self.tensordict = _tensordict
    tc.__dict__["tensordict"] = tensordict
    return tc
```

Reply:
`init` here is the dataclass' `__init__`?

I see where you're doing with this, just a bit awkward because currently the internals of `_from_tensordict` don't have access to `init`. We'll have to rewrite it as a wrapper of some sort.

Reply:
wouldn't calling `cls.__init__` be enough?

Reply:
But that's the tensorclass' init method right? So that's effectively what I'm currently doing anyway with

```python
tc = cls(**tensordict, batch_size=tensordict.batch_size)
```

Reply:
oh, my bad...in this PR we are still overwriting `cls` instead of using `datacls`.
you are right then, we need to do a wrapper like other functions